### PR TITLE
feat(resharding) - delete snapshot immediately after resharding is finished

### PR DESCRIPTION
**Pull request by wacban**
As long as the snapshot is present the storage overhead will keep increasing. This PR makes it so that the snapshot is deleted immediately after resharding is finished. The snapshot is only deleted if the node is not configured to take snapshots every epoch. 

Initially I wanted to put the trigger logic somewhere between sync jobs actor and chain but there it was hard to tell if resharding is finished for **all** the shards. It was made trickier by the fact that we only resharding shards that we will track next epoch and should only check those. At the end of the day I placed it in the state sync logic where it'll be triggered once all state syncs and hence all reshardings are done.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -3486,33 +3486,54 @@ impl Chain {
...
-                    .shard_uids()
-                    .collect();
-                let last_block = self.get_block(&head.last_block_hash)?;
-                let make_snapshot_callback = &snapshot_callbacks.make_snapshot_callback;
-                make_snapshot_callback(head.prev_block_hash, epoch_height, shard_uids, last_block);
-            } else if delete_snapshot {
-                let delete_snapshot_callback = &snapshot_callbacks.delete_snapshot_callback;
-                delete_snapshot_callback();
-            }
+        let Some(snapshot_callbacks) = &self.snapshot_callbacks else { return Ok(()) };
```
Comment (wacban) [inline chain/chain/src/chain.rs:3495]: Only a small refactoring in this method, should be noop.

**Comment by wacban** [comment]
cc @posvyatokum and @VanBarbascu who asked for this feature to be implemented
I had a look at logs from the resharding nayduck test and all seems ok. That being said in nayduck the resharding completes near instantly so it may not be the best test. The proper way to see it working would be to do a mocknet test. @posvyatokum since you're doing it anyway for the release, can you check if the snapshot is properly deleted at the right time? Or just point me to a mocknet host after the test is finished and I'll have a quick look at the logs.

**Comment by wacban** [comment]
Actually the integration tests cover this case. Still would be good to see it on a real db but even without it I'm fairly confident it should work.

File: chain/client/src/sync/state.rs [hunk 2]
```diff
@@ -702,6 +702,7 @@ impl StateSync {
         )?;
 
         if all_done {
+            chain.process_snapshot_after_resharding()?;
```
Comment (shreyan-gupta) [inline chain/client/src/sync/state.rs:705]: Would it make more sense to call this in `build_state_for_split_shards_postprocessing`? Why tag into the whole state sync process, which may run even if there's no resharding happening.

Comment (wacban) [reply chain/client/src/sync/state.rs:705]: yeah I wanted to but postprocessing is per shard

Comment (shreyan-gupta) [reply chain/client/src/sync/state.rs:705]: Oops, you're right! Apologies, ignore my comment!

**Pull request merged**